### PR TITLE
Add correct Content-Length on HEAD requests

### PR DESCRIPTION
**Pull request by jasontedor**
This commit fixes responses to HEAD requests so that the value of the
Content-Length is correct per the HTTP spec. Namely, the value of this
header should be equal to the Content-Length if the request were not a
HEAD request.

This commit also fixes a memory leak on HEAD requests to the main action
that arose from the bytes on a builder not being released due to them
being dropped on the floor to ensure that the response to the main
action did not have a body.

Relates #21077

### CODE REVIEW

File: distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java [hunk 1]
```diff
@@ -45,28 +45,31 @@ private void createTestDoc() throws UnsupportedEncodingException, IOException {
...
+        headTestCase("test", emptyMap(), 0);
+        headTestCase("test", singletonMap("pretty", "true"), 0);
     }
 
     public void testTypeExists() throws IOException {
         createTestDoc();
-        headTestCase("test/test", emptyMap());
-        headTestCase("test/test", singletonMap("pretty", "true"));
+        headTestCase("test/test", emptyMap(), 0);
+        headTestCase("test/test", singletonMap("pretty", "true"), 0);
```
Comment (nik9000) [inline distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java]: These aren't so much right as what we do now.

File: distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java [hunk 2]
```diff
@@ -45,28 +45,31 @@ private void createTestDoc() throws UnsupportedEncodingException, IOException {
...
     }
 
-    private void headTestCase(String url, Map<String, String> params) throws IOException {
+    private void headTestCase(String url, Map<String, String> params, int length) throws IOException {
         Response response = client().performRequest("HEAD", url, params);
         assertEquals(200, response.getStatusLine().getStatusCode());
         /* Check that the content-length header is always 0. This isn't what we should be doing in the long run but it is what we expect
```
Comment (nik9000) [inline distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java]: This comment is out of date now.

File: distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java [hunk 3]
```diff
@@ -45,28 +45,31 @@ private void createTestDoc() throws UnsupportedEncodingException, IOException {
...
          * that we are *actually* doing. */
-        assertEquals("We expect HEAD requests to have 0 Content-Length but " + url + " didn't", "0", response.getHeader("Content-Length"));
+        assertEquals(
+            "We expect HEAD requests to have 0 Content-Length but " + url + " didn't",
```
Comment (nik9000) [inline distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java]: The `0` in this is wrong now.

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by nik9000** [comment]
I think it'd be nice to move the test into the netty3 and netty4 modules because it is now them doing the work.

File: distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java [hunk 4]
```diff
@@ -45,28 +45,29 @@ private void createTestDoc() throws UnsupportedEncodingException, IOException {
...
 
-    private void headTestCase(String url, Map<String, String> params) throws IOException {
+    private void headTestCase(String url, Map<String, String> params, int length) throws IOException {
         Response response = client().performRequest("HEAD", url, params);
         assertEquals(200, response.getStatusLine().getStatusCode());
-        /* Check that the content-length header is always 0. This isn't what we should be doing in the long run but it is what we expect
-         * that we are *actually* doing. */
-        assertEquals("We expect HEAD requests to have 0 Content-Length but " + url + " didn't", "0", response.getHeader("Content-Length"));
+        assertEquals(
+            "We expect HEAD requests to have " + length + " + Content-Length but " + url + " didn't",
```
Comment (abeyad) [inline distribution/integ-test-zip/src/test/java/org/elasticsearch/test/rest/HeadBodyIsEmptyIT.java]: extra `+` in the message... ("+ Content-Length" -> " Content-Length")

**Review by nik9000 — APPROVED** [review]

**Review by jaymode — APPROVED** [review]

**Comment by abeyad** [comment]
LGTM2

**Comment by jasontedor** [comment]
Thanks @abeyad, @nik9000, and @jaymode.

**Pull request merged**